pallets/xyk/src/lib.rs

```diff
@@ -2307,29 +2426,63 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 			ExistenceRequirement::KeepAlive,
 		)?;
 
+		Pallet::<T>::deposit_event(Event::RewardsClaimed(user, liquidity_asset_id, mangata_amount));
+
 		Ok(())
 	}
 
 	fn promote_pool(liquidity_token_id: TokenId) -> DispatchResult {
 		ensure!(
-			!PoolPromotionStart::<T>::contains_key(liquidity_token_id),
+			<T as Config>::PoolPromoteApi::get_pool_rewards(liquidity_token_id).is_none(),
 			Error::<T>::PoolAlreadyPromoted,
 		);
-		let current_time = <frame_system::Pallet<T>>::block_number().saturated_into::<u32>() /
-			T::RewardsDistributionPeriod::get();
 
-		PoolPromotionStart::<T>::insert(&liquidity_token_id, current_time);
+		<T as Config>::PoolPromoteApi::promote_pool(liquidity_token_id);
 
-		let liquidity_assets_amount: Balance =
-			<T as Config>::Currency::total_issuance(liquidity_token_id.into()).into();
+		Pallet::<T>::deposit_event(Event::PoolPromoted(liquidity_token_id));
 
-		<T as Config>::PoolPromoteApi::promote_pool(liquidity_token_id);
+		Ok(())
+	}
 
-		LiquidityMiningPool::<T>::insert(
-			&liquidity_token_id,
-			(current_time, U256::from(0), U256::from(liquidity_assets_amount)),
+	fn activate_liquidity(
+		user: T::AccountId,
+		liquidity_asset_id: Self::CurrencyId,
+		amount: Self::Balance,
+	) -> DispatchResult {
+		ensure!(
+			<T as Config>::PoolPromoteApi::get_pool_rewards(liquidity_asset_id).is_some(),
+			Error::<T>::NotAPromotedPool
 		);
-		Pallet::<T>::deposit_event(Event::PoolPromoted(liquidity_token_id));
+		ensure!(
+			<T as Config>::Currency::free_balance(liquidity_asset_id.into(), &user).into() >=
+				amount,
+			Error::<T>::NotEnoughAssets
+		);
```

Comment:
This should be ensure_can_withdraw.
Frozen/locked tokens are not accounted for in free_balance.

Reply:
Please make similar changes across the pallet.
This is a major breaking issue

Reply:
Also you mentioned "somewhere I need “disposable” amount". Please make sure that this value is computed as free.saturating_sub(frozen)

Reply:
added free-frozen as disposable amount